### DISCUSSION TITLE
Compiling GO on Windows - dependency problems

### DISCUSSION THREAD

**Discussion by midimusic** [GrandOrgue/grandorgue#1145] [Help]
As an experiment I am trying to build/compile GO directly on Windows. I am configuring with the Windows version of Cmake-gui, which is known to work reliably. I am compiling DIRECTLY on Windows and am NOT using a Linux Cross compiler.

I tried using Mingw64 at first but configuration failed because the compiler does not contain pkg-config. I solved this by using Mingw32 instead as it does contain this package and is known to work properly.

I now have trouble with fftw3f - finding the package is easy both for 32 and 64 bits. However there is no recognised way to "install" this package on Windows and wherever I put it CMake does not recognise it. Firstly why does GO need it? - there is nothing in GO that uses fft's (fast fourier transforms) Secondly how should I locate the package so that CMake finds it i.e. where is cmake looking for it and what EXACTLY is it looking for? Putting the header and dll.s into the compiler's include and lib folders does not work - by the way this is where the pkg-config file is located.

CMake has no trouble finding wxWidgets and is happy to accept other missing minor dependencies such as gettext without creating an error and failing.

I suspect there will be similar problems with other minor dependencies such as wavepack which GO definitely needs. This is actually installed on the system so CMake should find it.

csw900

**Comment by midimusic** [comment]
Cmake did not find wavepack in spite of the executable being installed and useable. I suspect the source code is needed - but where to put it?

Further to my statement above I found other problems (I am using the source code V3.7.0)
Having commented out the requirement for fftw3f, zlib and wavepack the Configuration went through to completion and Generate was OK without further errors.
Compiling the resulting makefile produced many warnings and finally stopped with a fatal error. It appears to be unable to find the GO icon. I cannot find it either.

csw900

**Comment by oleg68** [comment]
@midimusic Lack of cmake-compatible dependency resolution on Windows is the most important reason why we use cross-compilation on Linux instead of a native build on Windows.

**Reply by midimusic** [reply]
Thanks oleg. Having investigated the lack of a GO icon, I have found the .svg file and the CMake instructions for converting it to a normal format using ImageMagick. Looking through the conversion scripts there is no option for Windows - thus there is no icon produced on Windows. It would have been just as easy to supply the already converted icons with options to select the correct one/s.

The annoying things are that Windows versions of all of these dependencies (including  ImageMagick) are easily available and downloadable. You are prioritising Apple and Linux/Unix over Windows while the combined market share for both is far less than that for Windows.

In my opinion it would be a good idea to build these dependencies into the source code for GO in the same way you have already done for Rtmidi etc. The source code for all of them is availabele and could easily be added by the user to a src/?? folder.

Anyway it was an interesting experiment, what a shame it did not turn out better.

csw900

**Reply by oleg68** [reply]
The main reason is a big technical complexity of adopting cmake for windows.

> In my opinion it would be a good idea to build these dependencies into the source code for GO in the same way you have already done for Rtmidi etc. 

I find a better idea is to get rid Rtmidi etc. from the source code and use the binary library dependencies. 

It is quite easy on Linux/Mac because they provide the `PkgConfig` dependency manager.

**Comment by midimusic** [comment]
Both Linux and Mac OS's are both basically Unix and are very similar, Windows is completely different and this is the problem.
Using library dependencies is always going to be easier - the library functions required by GO are all available for Windows but the problem is that cMake often does not find them. However if you put them all in some known location and tell it where they are then it should find them.

Pkg-config is also available for windows. There is a 64 bit lite version available which avoids gtk dependencies. (I have never used it however). I am also no expert on making cmake scripts.

I do know from my own experience how difficult it is to write code which will compile and work on all three systems. Bugs in wxWidgets do not help - StretchBlit only works properly on some systems and even the sample code provided does not always work correctly. Luckily GO avoids using this function.

csw900
